src/Moq/InvocationShape.cs

```diff
@@ -160,8 +161,29 @@ public bool Equals(InvocationShape other)
 				other.partiallyEvaluatedArguments = PartiallyEvaluateArguments(other.Arguments);
 			}
 
-			for (int i = 0, n = this.partiallyEvaluatedArguments.Length; i < n; ++i)
+			var lastParameter = this.Method.GetParameters().LastOrDefault();
+			var lastParameterIsParamArray = lastParameter != null && lastParameter.ParameterType.IsArray && lastParameter.IsDefined(typeof(ParamArrayAttribute));
+
+			for (int i = 0, li = this.partiallyEvaluatedArguments.Length - 1; i <= li; ++i)
 			{
+				// Special case for final `params` parameters, which need to be compared by structural equality,
+				// not array reference equality:
+				if (i == li && lastParameterIsParamArray)
+				{
+					if (this.Arguments[li] is NewArrayExpression e1 && other.Arguments[li] is NewArrayExpression e2 && e1.Expressions.Count == e2.Expressions.Count)
```

Comment:
I suspect we should really be comparing partially-evaluated arguments. The reason why this code accesses the original `Arguments` instead is to avoid having to deal with two distinct cases:

* `partiallyEvaluatedArguments[li]` may have been evaluated to a `ConstantExpression` wrapping an actual array&mdash;we'd no longer be able to use `ExpressionComparer` for its elements, but we'd have to do something akin to what `ConstantMatcher` does (i.e. comparing arrays using `SequenceEqual`).

* `partiallyEvaluatedArguments[li]` could have been left as a `NewArrayExpression`, meaning we'd use `ExpressionComparer` to compare its contained sub-`Expressions`.

❗️ Ideally, add more tests to see if the current approach is insufficient.